### PR TITLE
Add ability to rotate subscription keys for Text Analytics service

### PR DESCRIPTION
**Pull request by maririos**
Issue: https://github.com/Azure/azure-sdk-for-net/issues/8863

Following the patter done for Azure.Storage with [SharedKeyCredentials](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Common/src/StorageSharedKeyCredential.cs).

To keep this PR simple, test is not configure for recording. Plan to address it when we work on https://github.com/Azure/azure-sdk-for-net/issues/9091

### CODE REVIEW

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Let's think about naming on this.  It's a cross-cognitive concept, so we'll want to make sure it sounds nice and official.

Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Could we make this a separate class as well, the way storage does with SharedKeyCredential?

**Comment by annelo-msft** [comment]
It would be good to have a look at the [cognitive auth docs]( https://docs.microsoft.com/en-us/azure/cognitive-services/authentication) and do a little forward thinking about whether we would also want to support the Subscription-Region header, and if anything different would need to happen for single-service and multi-service subscription keys.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (kristapratico) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: I'm wondering if we should follow storage's lead here by making the credential a separate class, or if we should try to keep authentication by subscription key simple with just having the user provide a string. Am I missing a reason for why it should be a separate class?

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Because other cognitive services use the same auth methods, better to make it a separate class and then see how we make it in a shared way (shared package, etc) that other services will consume it too.

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: ServiceSubscriptionKey?

**Comment by maririos** [comment]
@annelo-msft 

>support the Subscription-Region header

Currently it is only used for ` Translator Text` service. We could ask the service team if this is intended to expand to other services (doesn't look like it) for it to be worth making general. 
Otherwise it could be something specific to that service.

>anything different would need to happen for single-service and multi-service subscription keys

Apart from pricing model, I don't see how our calls will be affected. According to docs it is mandatory to specify the region in the endpoint for multi service. It is ok to have it for single service too, and at the end is just an endpoint that the user gets from either the portal or the Azure CLI, so it will be included.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (kristapratico) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Adding more name possibilities: SubscriptionKeyCredential, APIKeyCredential. 

Just noticing that the portal refers to the key as both subscription key and API key in single-service accounts, and only as subscription key for multi-service accounts.

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: I like `SubscriptionKeyCredential` it helps to understand what to use it for, and as you point it out, works for both scenarios.

Comment (annelo-msft) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: I like that one too ... I've also spent more time thinking about the alternatives, including ApiSubscriptionKey, and I'm not as opposed to just SubscriptionKey as I initially was.  If there's a reason that we can't use the word Credential in something that's not in Azure.Identity, then SubscriptionKey may be the next best option.  I'll ask for Krzysztof's perspective.

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: >use the word Credential in something that's not in Azure.Identity?

@schaabs

Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: adding @KrzysztofCwalina

Comment (KrzysztofCwalina) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: @schaabs, could to recommend a good name here?

**Comment by maririos** [comment]
PR is on hold as the general guidelines for rotation keys are under discussion with Architecture Board.
Issue where discussion is happening: https://github.com/Azure/azure-sdk/issues/947

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (schaabs) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: A couple of thoughts. If this type were to remain in this package I would say `TextAnalyticsSubscriptionKeyCredential`. While it's quite wordy it does make it quite clear what clients this credential applies to. If you wanted to shorten this you could use one of the other variant names the documentation gives this key, such as `TextAnalyticsApiCredential`, or even simply `TextAnalyticsKeyCredential`. 

However, reading over this [Cognitive Services Authentication Doc](https://docs.microsoft.com/en-us/azure/cognitive-services/authentication), it seems like this credential type would likely apply to subscription key authentication across all cognitive services. If this is the case I would suggest moving this to a shared assembly so the same credential type can be used across all the service clients in `Azure.AI` namespace. In this case I would call it replace the "TextAnalytics" portion of the names with "CognativeServices", giving `CognativeServicesSubscriptionKeyCredential`, `CognativeServicesApiKeyCredential`, or `CognativeServicesKeyCredential`. 

Or if you prefer "AI" instead, `AISubscriptionKeyCredential`, `AIApiKeyCredential`, or `AIKeyCredential`. I'm not sure of the casing here. I believe the guideline is two letter acronyms are supposed to be all uppercase. But in some fonts it kind of reads like "AL Key Credential".

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Let's only add this header if we're using subscription key auth ... it's not relevant to AAD.  (Same on line 1747 below)

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/ServiceSubscriptionKey.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/ServiceSubscriptionKey.cs]: Per the updated guidelines, let's call this TextAnalyticsSubscriptionKeyCredential.  I don't see it referred to as API key anywhere in the docs, and it is heavily referred to as subscription key.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/ServiceSubscriptionKey.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/ServiceSubscriptionKey.cs]: I like Update a little better as it indicates you're changing an existing value rather than setting it for the first time.  We'll go with whatever the guidelines say once they're clarified.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: credential should be singular.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Let's also add the overload that takes a string subscriptionKey and instantiates a TextAnalyticsSubscriptionKeyCredential from it.  The arch board can tell us if they don't like it in our reviews with them, but the service team has requested that we make this as simple as possible for customers.  We'll want docstrings that make it clear that if you use that constructor, you won't be able to rotate credentials.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/samples/Sample5_RecognizePiiEntitiesBatch.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/samples/Sample5_RecognizePiiEntitiesBatch.cs]: Credential should be singular.  You could also new this inside the client constructor to avoid the variable name altogether.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (annelo-msft) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: I just saw that Krista and Johan decided not to add this for Python.  Let's understand why, and if their thinking applies here, let's follow suit.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Issue https://github.com/Azure/azure-sdk-for-net/issues/9558 created for that

**Comment by maririos** [comment]
@schaabs PR has been updated after comments for the Arch board meeting.
Note that instead of `Refresh` we are using `UpdateCredential` because we believe it makes clear the intention of the method.
If the final decision is to use `Refresh`, we'll adjust

**Comment by schaabs** [comment]
@maririos I'm creating the PR for the final guidelines now. In regards to your question about `Refresh` vs. `UpdateCredential`, from the arch board discussion I think we had consensus that `Refresh` was not the correct verb, and I do like `Update` better. However, what do you think about simply `Update` instead of `UpdateCredential`? Or alternatively `UpdateKey` since that's the only info on the credential you're updating?

**Comment by xirzec** [comment]
+1 for `UpdateKey`

**Comment by maririos** [comment]
@schaabs 
>what do you think about simply Update instead of UpdateCredential? Or alternatively UpdateKey since that's the only info on the credential you're updating?

For Text Analytics, I like `UpdateKey` too, but for something more generic, `Update` is good. :|
Naming is so hard...

File: sdk/textanalytics/Azure.AI.TextAnalytics/README.md
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/README.md]: Depending on where we land with the constructors -- did we decide to keep the string constructors? -- let's either put the string constructor here with a comment that you can roll a credential with another constructor, or put snippets for both and call out the difference.

File: sdk/textanalytics/Azure.AI.TextAnalytics/samples/Sample2_AnalyzeSentiment.md
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/samples/Sample2_AnalyzeSentiment.md]: For these, let's say you'll use a .. subscription key **credential**.  And let's add a short sentence saying something like "passing in a TextAnalyticsSubscriptionKeyCredential" allows you to update credentials without creating a new client.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: Let's add the same sentence here as in the samples ... "Passing ... allows you to ...."

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsSubscriptionKeyCredential.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsSubscriptionKeyCredential.cs]: Let's add the comment here about why you would use this class.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsSubscriptionKeyCredential.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsSubscriptionKeyCredential.cs]: Against the TextAnalytics service.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsSubscriptionKeyCredential.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsSubscriptionKeyCredential.cs]: to use to authenticate with the service.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/AuthenticationTests.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/AuthenticationTests.cs]: I think we should either add this to Live tests or name it Live tests so it's clear.

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/samples/SampleSnippets.cs
Comment (annelo-msft) [inline sdk/textanalytics/Azure.AI.TextAnalytics/tests/samples/SampleSnippets.cs]: Shouldn't this stay subscriptionKey?

**Review by annelo-msft — APPROVED** [review]
Some small comments.

File: sdk/textanalytics/Azure.AI.TextAnalytics/samples/Sample2_AnalyzeSentiment.md
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/samples/Sample2_AnalyzeSentiment.md]: Ohh nice catch. I didn't realized this was here. 
I will modify

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/AuthenticationTests.cs
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/AuthenticationTests.cs]: Fair. Once live test pr is merged, I will create a PR to move this test there, for now I will leave it here

File: sdk/textanalytics/Azure.AI.TextAnalytics/README.md
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/README.md]: For now I'm going with 4 constructors. 
After UX study, if wanted, we can add the string constructor

File: sdk/textanalytics/Azure.AI.TextAnalytics/tests/AuthenticationTests.cs
Comment (maririos) [reply sdk/textanalytics/Azure.AI.TextAnalytics/tests/AuthenticationTests.cs]: Comment Added to https://github.com/azure/azure-sdk-for-net/issues/9091

**Comment by maririos** [comment]
Merging for now. If guidelines suggest other naming, I will create a new PR with that.

**Pull request merged**
